pkg/kubemark/controller.go

```diff
@@ -0,0 +1,353 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubemark
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/informers"
+	informersv1 "k8s.io/client-go/informers/core/v1"
+	kubeclient "k8s.io/client-go/kubernetes"
+	listersv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/golang/glog"
+)
+
+const (
+	namespaceKubemark = "kubemark"
+	hollowNodeName    = "hollow-node"
+	nodeGroupLabel    = "autoscaling.k8s.io/nodegroup"
+)
+
+// Controller is a simplified version of cloud provider for kubemark. It allows
+// to add and delete nodes from a kubemark cluster and introduces nodegroups
+// by applying labels to the kubemark's hollow-nodes.
+type Controller struct {
+	nodeTemplate    *apiv1.ReplicationController
+	externalCluster externalCluster
+	kubemarkCluster kubemarkCluster
+}
+
+// externalCluster is used to communicate with the external cluster that hosts
+// kubemark, in order to be able to list, create and delete hollow nodes
+// by manipulating the replication controllers.
+type externalCluster struct {
+	rcLister        cache.SharedInformer
+	rcListerSynced  cache.InformerSynced
```

Comment:
s/rcListerSynced/rcSynced (it's not the Lister that is synced, but the informer)
Also make the similar change elsewhere.